### ISSUE TITLE
Implementation of Configuration Management - Part 2

### ISSUE THREAD

**Issue by esaber76** [US-EPA-CAMD/PI-1#6425]
This is a follow on to #5995 to fix any remaining critical issues with the new Configuration Management implementation.

Current issues to address:

1. For units that never submitted a MP, receiving an "Error loading summary of configuration changes." message on the Change Summary popup after saving.

2. Existing stack/pipe configurations:
    - After ending a stack/pipe configuration, if I revert the new inactive configuration and then the active configurations, the new active MP's that were created still exist in camdecmpswks.monitor_plan. They also show up with a null Configuration column on the Select Configurations screen. Selecting one of these results in a "Cannot read properties of null (reading '1')" error. **Note:** if I revert the active MP's first and then the inactive, there are no problems. The revert order is also not a problem when adding new stack/pipe configurations. 
    - When new records are added to camdecmpswks.monitor_plan_reporting_freq, upate_date is also set. This should be null unless a subsequent change occurs.
    - Active MP's need to be submitted before an inactive MP, or they all need to be submitted at once. If I try to submit only the inactive MP first, the MP is stuck in a loop where the status just remains resubmission required over and over.
        - ECMPS 1.0 forces users to submit the inactive MP prior to being able to submit the active MP's. We can split this out into a different ticket if needed.
    - When attempting to end stack/pipe and USC records before the last day of the quarter for the latest emission evaluation in the UI, I receive the "Error loading summary of configuration changes." message on the Change Summary popup after saving. **Note:** I do get the appropriate messages when doing this via file import.
    
3. Creating new stack/pipe configurations:
    - I'm still unable to add a new stack/pipe configuration for ORIS 544, 11 and 12. Attempted to add CP1, 11, 12 effective 10/01/2023 and receive the "Error loading summary of configuration changes." message on the Change Summary popup after saving. 
    - I was able to add a new stack/pipe configuration for ORIS 542, 10 and 11 effective 10/01/2024 but:
        - update_date is set when new records are added to the monitor_plan, stack_pipe, unit_stack_configuration, and monitor_plan_reporting_freq tables. This should be null unless subsequent changes occur.
        - Unable to complete the submissions for these configurations.
            - For the inactive MP submissions, I see the following error in camdecmpsaux.submission_set: 
`{"response":{"message":"insert or update on table \"monitor_plan_location\" violates foreign key constraint \"fk_monitor_plan_location_monitor_plan\"","statusCode":500},"status":500,"message":"insert or update on table \"monitor_plan_location\" violates foreign key constraint \"fk_monitor_plan_location_monitor_plan\"","name":"EaseyException","metadata":{}}`
            - For the active MP submission, I see the following error in camdecmpsaux.submission_set:
`{"response":{"message":"insert or update on table \"monitor_location\" violates foreign key constraint \"fk_monitor_location_stack_pipe\"","statusCode":500},"status":500,"message":"insert or update on table \"monitor_location\" violates foreign key constraint \"fk_monitor_location_stack_pipe\"","name":"EaseyException","metadata":{}}`
    - SIPNOX only examples:
        - ORIS 50607, RSB1 and RSB2. Adding new CP stack/pipe record and associated USC records effective 10/01/2024.
            - ~~RSB2 (single unit configuration) is showing under New Plans instead of Ended Plans on the Change Summary screen. Saving the changes results in an inactive monitoring plan for RSB2 (correct), a new monitoring plan for RSB2 (incorrect), and a new active CP configuration with RSB2 (correct).~~ This is due to the RSB2 Monitor Plan begin date not aligning with the unit's begin date (the begin date of its earliest monitor method), and will not be addressed in this ticket.
            - ~~The Reporting Frequency indicates Ozone Season for the new plans created. SIPNOX-only units have the option to report annually or during the ozone season. If the ended plan is annual in this case, the new plan(s) should be annual. If the ended plan is ozone season in this case, the new plan(s) should be ozone season. **Note:** ORIS 54638, CS1, 040, 041 is an example of a stack/pipe configuration with SIPNOX units that report only in the ozone season. Ending this configuration creates annual reporting frequency records for the new plans in error. We can discuss this more with Dwayne. ECMPS 1.0 appears to set all new plans (adding or ending stack/pipe configurations) to annual for SIPNOX-only units. We should at least be consistent with whichever path we choose.~~ Leaving SIPNOX only locations as is after further discussions as the os_ind in camdmd.program_code is 1. The ORIS 54638 example is also RGGI affected and the reason the records were set to annual. We can either adjust as needed with scripts or hopefully CSA still.
            - ~~**Note:** Unit 24 gets ended when these changes occur, which is sort of odd as it wasn't targeted and it's retired. ECMPS 1.0 also ends this plan with the same type of changes, but the end dates are different. ECMPS ends the plan 2023 Q1 (using last emissions evaluation?). 2.0 ends the plan with the retired status (2016 Q4). This is an odd situation where non-operating files were submitted for years and then the unit was retired way after the fact. We should discuss this with Dwayne and maybe others to determine what to do for this situation since EM data was submitted through 2023 Q1.~~ Leaving as is after further discussions. We can fix with scripts if this were ever to occur and problems surfaced.
    - Receiving the "Error loading summary of configuration changes." message on the Change Summary popup after saving when adding USC begin dates that are before the quarter of the last emission report in the UI. I "think" this was actually working before the last round of changes.
    - Receiving the "Error loading summary of configuration changes." message on the Change Summary popup after saving when attempting to save stack/pipe or USC records that have end dates before the begin dates in the UI.

4. Also see issues noted on #6434.

**Comment by maxdiebold-erg** [comment]
@esaber76 For the first bullet point of number 2, this behavior is occurring due to a constraint on _camdecmpswks.monitor_plan_location_ that drops rows when the associated monitor location is dropped (which occurs during the first step of reverting a monitor plan). These rows may or may not be associated with the MP that is being reverted. In the second step, when copying back from official to the workspace, all _monitor_plan_location_ records associated with the plan's locations are copied into the workspace (again, even those not affiliated with the target MP). This has worked fine for plans that already exist in Official, but clearly it won't restore the MPL records for a peripheral plan that isn't yet in Official.

**EDIT:** This would be fixed by the changes outlined by #6266, particularly:
>Skip deleting a MONITOR_LOCATION row if it is associated with a MP in the workspace other than the MP being reverted

To allow for testing this issue, the necessary changes have been implemented in https://github.com/US-EPA-CAMD/easey-db-scripts/pull/70.

**Comment by maxdiebold-erg** [comment]
I believe the "submission loop" issue is related to the other submission issues noted. I will make the changes to fix that issue, but can we split the explicit ordering into a different ticket?

**Comment by esaber76** [comment]
1. We need to better handle updating the begin_rpt_period_id in the monitor_plan and monitor_plan_reporting_freq tables. Currently, if a user creates a new MP in Configuration Management, begin_rpt_period_id is set to the current quarter based on the current date. If the user then adds methods that begin before or after (this is blocked by #6534 but assume the behavior is the same) this quarter, begin_rpt_period_id is not updated for the MP. This can result in a duplicate MP getting created. One based on the initial Configuration Management setting and another one that aligns with the method begin quarter.
2. ~~Internal Server Errors are getting returned when reverting some MP's. Per Dwayne, copy_monitor_plan_to_workspace does not currently include the mats_method_data table. Examples: ORIS 1710, 1, 2, and 3; CS0009, 1, 2; MS3A, MS3B, 3. Also experienced this problem for ORIS 1012, CS023, MS2, 2, 3 after importing a json file that retired MS2 and the associated USC record and then attempting reverts on the ended and new MP's.~~ **Addressed in #6266**
3. Internal Server Error when attempting to add a new stack/pipe and USC records via json import. Attached example for ORIS 4940, CP1 (new), 1503, 1504. 
[ORIS 4940 CP1 1503 1504 New MP.json](https://github.com/user-attachments/files/18363870/ORIS.4940.CP1.1503.1504.New.MP.json)
4. ~~The error message below is shown when attempting to submit new active MP's prior to newly ended MP's. We discussed using this message instead. "Inactive monitoring plans for at least one of the locations in the current monitoring plan need to be submitted prior to submitting the current, active monitoring plan." This change was likely overwritten by other updates to the queuing/ submission process. We can live with this for now if it's not a simple change and spin off a new ticket if needed.~~ **Moved to #6546**

**Comment by maxdiebold-erg** [comment]
A note about point number 3 described above:
> Internal Server Error when attempting to add a new stack/pipe and USC records via json import. Attached example for ORIS 4940, CP1 (new), 1503, 1504. 

It turns out that this issue was caused by calls to reset an MP to "needs eval" while importing **components** and **monitor loads**, even if the MP had not yet been created. For all other types of monitoring location data, the method that resets an MP to "needs eval" is only called **if the create or update is not occurring as part of an import**. For imports, the plan is reset where necessary(?) by the top-level import function.

To fix the issue and to be consistent, I added the "is import" flag to the aforementioned types of monitoring location data. However, there are now cases (e.g. ORIS 4940 with the JSON file provided [above](https://github.com/US-EPA-CAMD/easey-ui/issues/6425#issuecomment-2580632730)) where importing a plan does **not** reset the plan to "needs eval", even if location data has changed.

**Question:** Should importing an MP _always_ reset the associated plan(s)?

**Comment by esaber76** [comment]
> A note about point number 3 described above:
> 
> > Internal Server Error when attempting to add a new stack/pipe and USC records via json import. Attached example for ORIS 4940, CP1 (new), 1503, 1504.
> 
> It turns out that this issue was caused by calls to reset an MP to "needs eval" while importing **components** and **monitor loads**, even if the MP had not yet been created. For all other types of monitoring location data, the method that resets an MP to "needs eval" is only called **if the create or update is not occurring as part of an import**. For imports, the plan is reset where necessary(?) by the top-level import function.
> 
> To fix the issue and to be consistent, I added the "is import" flag to the aforementioned types of monitoring location data. However, there are now cases (e.g. ORIS 4940 with the JSON file provided [above](https://github.com/US-EPA-CAMD/easey-ui/issues/6425#issuecomment-2580632730)) where importing a plan does **not** reset the plan to "needs eval", even if location data has changed.
> 
> **Question:** Should importing an MP _always_ reset the associated plan(s)?

@maxdiebold-erg by "always reset the associated plan(s)", are you referring to setting needs eval to Y? My experience with ECMPS 2.0 thus far has been that any MP import (whether data was changed or not) sets the plan to needs evaluation.

**Comment by mark-hayward-erg** [comment]
@esaber76 Test has been updated with the latest SQL changes for #6266.

**Comment by esaber76** [comment]
Per previous discussions, moving this ticket and the functionality currently in place to Ready for Deployment. Several new tickets were created to address remaining issues. Some of these were not directly related to the implementation of Configuration Management. They are pre-existing issues found during this testing effort.  

- #6437 
- #6438 
- #6533 
- #6534 
- #6535 
- #6536 
- #6537 
- #6593 
- #6594 
- #6597
